### DISCUSSION TITLE
STM8S003 Uart temp stability

### DISCUSSION THREAD

**Discussion by sbridger** [TG9541/stm8ef#446] [General]
Do you have any idea of how well the uart works without a crystal vs temperature?
Is your solar controller outside or inside, and have you had any issues?

**Comment by TG9541** [comment]
The 16MHz HSI is remarkably stable - and I have never run into any problems when using a host UART with standard (crystal controlled) accuracy. It depends, of course on the temperature range that you're targeting. 


![image](https://user-images.githubusercontent.com/5466977/171501261-ca9f9378-8270-4ac1-8605-f9f75c8d74dc.png)

My solar controller is inside, but I think, as long as your application won't have to work in a really cold environment it won't matter too much.

If a stable phase reserve is really important for your UART communication you can still look into LIN style baud rate tuning techniques.

**Comment by sbridger** [comment]
Thanks, I was interested in any adverse experiences vs theory.
It's to go at the top of an antenna mast as a uart->tuner relays adaptor, which is a pretty trivial job.
Not very cold here.

It has to go into halt between commands for interference reasons.
The tricky part is I am trying to work out a way to get bidirectional data, and 400mA of power down the same coax. 
I don't think the STM8 uarts have any modulation modes (eg 38kHz ASK) like PICs do? e.g timer gated by uart.
I suppose that a software send-only uart might be the answer, since it is the back channel alone that is the problem, and since it has nothing else to do when sending.

**Reply by TG9541** [reply]
No, not to my knowledge. The STM8L family has an IR remote control feature for gated pulse generation (but I didn't need it for [IR remote control](https://gist.github.com/TG9541/40b811c3a611eb02ba9e3693f99222ac)).

The STM8 UART IrDA mode might help to implement bidirectional communication with a modulated signal. Generating and decoding a modulation is yet a different story.

I guess that there is a lot of HAM lore in this area (e.g., use FM and a 4046 PLL for decoding?) - how would you approach that?
